github/commitstatus.go

```diff
@@ -129,197 +120,71 @@ func (s CommitStatus) Add(sha, state, targetURL, description string) error {
-	client := &http.Client{Timeout: time.Second * 30}
-	resp, err := client.Do(req)
-	if err != nil {
-		return httpResponse{}, fmt.Errorf("http client Do: %w", err)
+		body, _ := io.ReadAll(resp.Body)
+		return NewGitHubError(resp, errors.New(strings.TrimSpace(string(body))))
 	}
-	defer resp.Body.Close()
-
-	response.statusCode = resp.StatusCode
-	respBody, _ := io.ReadAll(resp.Body)
-	response.body = strings.TrimSpace(string(respBody))
 
-	// GH API BUG
-	// According to
-	// https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/
-	// each reply to a GH API action will return these entries in the header:
-	//
-	// X-Accepted-OAuth-Scopes:  Lists the scopes that the action checks for.
-	// X-OAuth-Scopes:           Lists the scopes your token has authorized.
-	//
-	// But the API action we are using here: POST /repos/:owner/:repo/statuses/:sha
-	//
-	// returns an empty list for X-Accepted-Oauth-Scopes, while the API documentation
-	// https://developer.github.com/v3/repos/statuses/ says:
-	//
-	//     Note that the repo:status OAuth scope grants targeted access to statuses
-	//     without also granting access to repository code, while the repo scope grants
-	//     permission to code as well as statuses.
-	//
-	// So X-Accepted-Oauth-Scopes cannot be empty, because it is a privileged operation,
-	// and should be at least repo:status.
-	//
-	// Since we cannot use this information to detect configuration errors, for the time
-	// being we report it in the error message.
-	response.oauthInfo = fmt.Sprintf("X-Accepted-Oauth-Scopes: %v, X-Oauth-Scopes: %v",
-		resp.Header.Get("X-Accepted-Oauth-Scopes"), resp.Header.Get("X-Oauth-Scopes"))
-
-	// strconv.Atoi returns 0 in case of error, Get returns "" if empty.
-	response.rateLimitRemaining, _ = strconv.Atoi(resp.Header.Get("X-RateLimit-Remaining"))
-
-	// strconv.Atoi returns 0 in case of error, Get returns "" if empty.
-	limit, _ := strconv.Atoi(resp.Header.Get("X-RateLimit-Reset"))
-	response.rateLimitReset = time.Unix(int64(limit), 0)
-
-	// time.RFC1123 is the format of the HTTP Date header
-	// example: Date:[ Mon, 02 Jan 2006 15:04:05 MST ]
-	// https://datatracker.ietf.org/doc/html/rfc2616#section-14.18
-	date, err := time.Parse(time.RFC1123, resp.Header.Get("Date"))
-	if err != nil {
-		return httpResponse{}, fmt.Errorf("failed to parse the date header: %s", err)
+	if err := cs.target.Retry.Do(Backoff, Classifier, workFn); err != nil {
+		return cs.explainError(err, state, sha, url)
 	}
-	response.date = date
 
-	return response, nil
+	return nil
 }
 
-func (s CommitStatus) checkStatus(resp httpResponse, state, sha, url string) error {
-	var hint string
-
-	switch resp.statusCode {
-	case http.StatusCreated:
-		// Happy path
-		return nil
-	case http.StatusNotFound:
-		hint = fmt.Sprintf(`one of the following happened:
+// TODO: can we merge (at least partially) this function in GitHubError.Error ?
```

Comment:
I see that you have similar questions I had...

I would very much like to get rid of this function/method or rewrite it nicer, but at the time I could not find any reasonable solution for it...